// @streamdal/node-sdk/cjs/streamdal.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Streamdal = exports.OperationType = exports.Audience = void 0;
const sp_common_1 = require("@streamdal/protos/protos/sp_common");
Object.defineProperty(exports, "Audience", { enumerable: true, get: function () { return sp_common_1.Audience; } });
Object.defineProperty(exports, "OperationType", { enumerable: true, get: function () { return sp_common_1.OperationType; } });
const uuid_1 = require("uuid");
const audience_js_1 = require("./internal/audience.js");
const grpc_js_1 = require("./internal/grpc.js");
const heartbeat_js_1 = require("./internal/heartbeat.js");
const metrics_js_1 = require("./internal/metrics.js");
const process_js_1 = require("./internal/process.js");
const register_js_1 = require("./internal/register.js");
class Streamdal {
    constructor({ streamdalUrl, streamdalToken, serviceName, pipelineTimeout, stepTimeout, dryRun, audiences, }) {
        var _a, _b;
        if (process.env.NODE_ENV === "production") {
            console.debug = () => null;
        }
        const url = streamdalUrl ? streamdalUrl : process.env.STREAMDAL_URL;
        const token = streamdalToken ? streamdalToken : process.env.STREAMDAL_TOKEN;
        const name = serviceName ? serviceName : process.env.STREAMDAL_SERVICE_NAME;
        if (!url || !token || !name) {
            throw new Error(`Required configs are missing. You must provide configs streamdalUrl, streamdalToken and serviceName 
        either as constructor arguments to Streamdal() or as environment variables in the form of STREAMDAL_URL, STREAMDAL_TOKEN and STREAMDAL_SERVICE_NAME`);
        }
        const sessionId = (0, uuid_1.v4)();
        const grpcClient = (0, grpc_js_1.client)(url);
        this.configs = {
            grpcClient,
            sessionId,
            streamdalUrl: url,
            streamdalToken: token,
            serviceName: name,
            pipelineTimeout: (_a = pipelineTimeout !== null && pipelineTimeout !== void 0 ? pipelineTimeout : process.env.STREAMDAL_PIPELINE_TIMEOUT) !== null && _a !== void 0 ? _a : "100",
            stepTimeout: (_b = stepTimeout !== null && stepTimeout !== void 0 ? stepTimeout : process.env.STREAMDAL_STEP_TIMEOUT) !== null && _b !== void 0 ? _b : "10",
            dryRun: dryRun !== null && dryRun !== void 0 ? dryRun : !!process.env.STREAMDAL_DRY_RUN,
            audiences,
        };
        // Heartbeat is obsolete
        setInterval(() => {
            void (0, heartbeat_js_1.heartbeat)(this.configs);
        }, heartbeat_js_1.HEARTBEAT_INTERVAL);
        setInterval(() => {
            void (0, metrics_js_1.sendMetrics)(this.configs);
        }, metrics_js_1.METRIC_INTERVAL);
        void (0, audience_js_1.addAudiences)(this.configs);
        void (0, register_js_1.register)(this.configs);
    }
    processPipeline({ audience, data, }) {
        return __awaiter(this, void 0, void 0, function* () {
            return (0, process_js_1.retryProcessPipeline)({ configs: this.configs, audience, data });
        });
    }
}
exports.Streamdal = Streamdal;
